### ISSUE TITLE
(req) Ssh for Making auth users or codes

### ISSUE THREAD

**Issue by Pok4** [openNDS/openNDS#417]
Great module!
I need to be implemented some new functionality if is possible...
https://forum.openwrt.org/t/wifi-for-hotel-with-20-floors/145396/1 (read first post)
I will use open wrt and this one.
I want with ssh to create users/passwords or other type of authentication (codes) with expiration period with ssh.

I develop a portal based on php which works with php ssh library and i want to touch the device from there.
So.. I need a ssh functionality for this module.
For creating/deleting/set expiration date for guests.

If you need more info - im here.
If this is possible will be great feature.

I want to catch logs for visited browser links for specific guest too - with ssh. But i dont think that is related to this module.

Greetings from Bulgaria!

**Comment by bluewavenet** [comment]
@Pok4 
A good starting point is to read the documentation at:
https://opennds.readthedocs.io/

OpenNDS **_MUST_** be running on the router that provided the ip subnet gateway, it cannot be run on an Access Point.

This is a requirement for **_ALL_** captive portals to comply with the requirements of CPD/CPI to give an automatic popup of the login pages.

You will see that there are two parts to a captive portal:
1. The part that does the capturing
2. The portal pages used by the client to log in (known as the FAS, or Forwarding Authentication Service)

In openNDS, the portal (FAS) pages can be:
 * local to the openNDS router and served by the built in openNDS webserver
 * local to the openNDS router and served by a third party web server
 * remote from the openNDS router and served by a third party web server on the same local network
 * remote from the openNDS router and served by a third party web server hosted in the Internet.
 

> Ssh for Making auth users or codes

Why ssh? Sure, ssh access to the router for remote maintenance is a good idea, but the user account details will be held on the FAS server. Most installations implement an autonomous user account generation based on the user login form information. Anything else will become vastly expensive for the captive portal venue (in this case your hotel).

If you have 140 rooms with one person staying one night, then that is 140 user accounts to be created per day, at any time during the day. If this process was not autonomous, you would have to supply teams of support staff available 24 hours a day.. non trivial and very costly to the venue.

In comparison, an autonomous system with user driven account generation will not require any staff.

> I want to catch logs for visited browser links for specific guest too

**This is not legal** in most countries, particularly within the EU. It is related to openNDS as you would implement it on the openNDS router. We can discuss the legal details here if you wish, and indeed you should be fully aware of the issues if you are serious about implementing a venue captive portal, either a one off, or as a business supplying the service to venues.

You have clearly put in a great deal of thought into this, which is good, and you see it as a business opportunity.
However there are many misconceptions that you may have and lack of experience in the field makes you a little naive.
The fact that you are asking, and have posted here and in the forums is very much in favour of you possibly making a success of it at least technically.

The business model will also need a lot of thought, for example, the days of monetising Internet access at guest venues is a thing of the past, apart from very short term opportunities in some countries with very underdeveloped Internet infrastructure. Free Internet access is expected by all users as a necessary service supplied by venues. If it is not free, then users will choose a different venue. From the point of view of your business, revenue will come from installation and ongoing maintenance, and not from charging users for access.

**Comment by Pok4** [comment]
Thanks for explain. I open the docs and i use the search first. I type ssh and i not found what i need.
I have basic knowledge about captive portal (I search articles Last days)

Yes, I will have a router with open wrt and multiple access points on the diffetent floors.
With this scenario its ok?
I will configure the router and i Will connect the access points. (for now i look at ubiquiti and ruckus) i will decide later, because we Will visited the hotel and we Will expect previous isp devices. And we Will talk with the owners. 

I prefer to work with ssh. I think that ssh supports certs for login (I dont think is a bad or unsecurely type of connection) 

I work with phpseclib in the past and its fast and easy to use. Now its a bit old. I can use built in library:
https://www.php.net/manual/en/function.ssh2-connect.php
https://www.php.net/manual/en/function.ssh2-exec.php

Of course maybe a frame with this (capsula) but for now i not search about this. 

If there is a variant without ssh i can work with this variant too. (but the connection must be supporting from php and to send  the commands with php too) 
In the past i use sockets (if there is variant with sockets its not a problem) 
As i said i can work with other type of connection but must be supported from php. 


About for the login page - this 4 options its ok for me. Login page can be in router web Server - its ok. 
I need only to create/delete and set expire date for users. 

About gdpr and the history of logs (visited links) - In our country this law exists but i not read about it. (im a young man and this stuff is not in my interests) 
This will be discussed when we start the project (with clients, my parents and our lawyer and etc..) 
I Just ask a questions which will be important for
The project.
If this is not ok to you - Just ignore it.

If other things is possible - i Will be glad to use it on my router which is on courriers from Sweden to Bulgaria now.
I will not ask the client for money for nothing. They only need to have internet connection on their own. (or on our firm name but with specific Contract) My web service and my equipment must be liked from them. I think that they have old equipment, bad quality and much more investitions which they pay monthly.
They will save money with my system and they de ide how to act about their clientes.
My Commission will be 0,50 euro cents from a client per day.
I will track every thing in my php web system and i  Will know every thing which is needed to pay from my clientes to me.
It will have billing system and sortable system, hronology and much much more (I can extended the system with new ideas in the Future)

For now this is the important thing for me. Without this - the project will not happen.
If is not possible here, maybe i can look around for other os for router which i order (teklager) that supports captive portal and i will start a topics in their communities.

Si if you need a more info - please write.
Sorry for my english.

**Comment by Pok4** [comment]
Yesterday i look for phpseclib and visit sourceforge site and there is a old info.
Now i spend some time and i see that they still update the library.
https://github.com/phpseclib/phpseclib

There is another variant - https://github.com/faraweilyas/sshbunny

It depends on developer Needs.

I can use standalone ssh library in php too.
This is de ide by developer. They Support composer and can be integrated on custom framework or another famous ones.
I have custom framework which is very easy to use and understand.
I think there is a lot more ssh based libraries but i dont want to waste my time for research. This two is enough.

**Comment by Pok4** [comment]
I want to be easy for you and i search a thing - https://stackoverflow.com/questions/10823269/is-it-possible-to-upload-files-using-ssh-connection

If you make the module to can read txt file locally on device and in this file is contain all users+passwords+expire date (epoch will be good) will be easy for me too...
Because i can read and modify the file with ssh too. Another option is to create ftp Server on device and to use sftp connection from phpseclib to read and modify the file. This will be workaround ofcourse.

This is variant to use flat file data base (txt file)

Some of open wrt Community tells about traffic limit, if is possible something like this:
Usr1:pass1:142764677:200mb
This will be:
Username
Password
Expire date in epoch time format
Bandwidth limit

If you can read the file and will be easy for you this way will be ok for me too.

I think i can setup ssh by default and your module is not ssh relared. I can touch directory with txt file and i Will make modification to the file.

**Comment by bluewavenet** [comment]
@Pok4 

> Because i can read and modify the file with ssh too. Another option is to create ftp Server on device and to use sftp connection from phpseclib to read and modify the file. 

**_You will not be able to access the openNDS router remotely_** unless you have either:
1. A fixed Internet ip address and set up port forwarding (a much more expensive Internet feed than the usual type, if it is even available to you), or
2. A NAT traversal tunnel of some kind to an Internet hosted server that you would use as a gateway (an Internet hosted service is required for this).

**_You do not need to create either of these as:_**
1. OpenNDS autonomously manages users for you (customisation is supported in the OpenNDS API for building FAS servers).
2. OpenNDS has it's own built in NAT traversal functionality when using an Internet hosted FAS server.

Building your own remote, ssh based management system is a pointless exercise as OpenNDS has everything built in to do what you need and more.
**_YOU DO NOT NEED SSH LIBRARIES_** and in fact this approach will not even work without a great deal of effort and additional expense of purchasing the correct type of Internet feed. This is not a function of OpenNDS, it is a function of how the Internet works.

**Comment by Pok4** [comment]
Maybe i not explain every thing.
I think that your openNDS is a module for openwrt?
If this is right, does openwrth have ssh functionality to touch whole device files and folders? Ddwrt have ssh in options and you can be enable or disabled and modified the settings. (this is by default)

Here in Bulgaria we using only static ips. (Just saying) i dont know a isp with dynamic ips (maybe in villages or small town)

I read a docs about ubiquiti and Other routers and they have built in captive portal with bunch of options, tracking and etc.. Even ruiji Networks offer more settings and cool Interface for web login page. Bunch of settings Really..

But this is not works for me. I want this options to be available from my script and offer them to my clientes via my website.

Every hotel will have user and password. The hotels will have different ssh data for their specific routers.
When hotels loggedin in my website they can create users and passwords for WiFi auth (for captive portal)
They have options to add custom fields like room, floor, name, leaving date and more settings.

I dont want to give to them router web Server and user and passwords to login directly in router and to touch the device settings.

The idea is not works in this way.

I want to buy hosting and domain.
My portal will have multilanguage Support, bunch of tracking, history and etc.. This will available to me and the hotels.
Online payment will be available too. (PayPal, paymentwall and stripe) maybe and easypay (works in Bulgaria)

I want to have custom site with this function.
It will be mobile friendly, I can make changable themes working with cookies and so on..

The idea is to not to need login in router settings. I want to build management system which working for hotels, offices, hospitality and Others.
For every client - different settings.
I want to build every thing from scratch.

In other words i purchase this router because is with good specs and works with open wrt.
Either way i guess i can buy ruiji or ubiquiti router with this features. But as i said they not Support such things. I already asked.
My Last options is a few os like open wrt, ddwrt and 2-3 more.

**Comment by bluewavenet** [comment]
@Pok4 

> I think that your openNDS is a module for openwrt?

It is a software package for Linux that is available as pre-cross-compiled binaries on OpenWrt.

> If this is right, does openwrt[] have ssh functionality to touch whole device files and folders?

Yes of course it does. But that does not mean you can access ssh on the router remotely. You may have a "static" ip, but is this an rfc1918 address? I suspect that it is, as blocks if Internet accessible ip addresses are extremely expensive if available at all, and any ISP would be insane to just give them away.
If you are getting an Internet accessible static ip address without paying a significant premium for it, then you are extremely lucky for now but you will most likely find it will be short lived.

> When hotels loggedin in my website they can create users and passwords for WiFi auth (for captive portal)

So every time someone checks in to the hotel, a member of staff will have to log in to your web site and create an account for the guest, then give the guest the username and password?
Good luck with that - this will take time to do and time costs money so the hotel will very quickly decide this is not a good idea and, simply put, guests will end up, pretty quickly, without Internet access.

>  They have options to add custom fields like room, floor, name, leaving date and more settings.

Sure, openNDS supports this.

> I want to buy hosting and domain.

Good. That is the best solution for large venues with openNDS. Each hosted service will be capable of handling many unique openNDS instances from many venues (your customers).

> My portal will have multilanguage Support

This is fully supported in a custom FAS - it is purely a matter of coding in a database if languages - all standard stuff.

> bunch of tracking, history and etc.. This will available to me and the hotels.

**_This is illegal_** in the European Union and most other countries. Both you and the hotel would be open to prosecution under EU laws.

> Online payment will be available too.

I thought you said you will not be charging guests? This might be acceptable to Russians, but guests from EU countries, the UK, the USA, Australia, etc etc will be very upset by this - they will all expect a FREE Internet service.

> I want to have custom site

Yes of course - that is what FAS is all about, creating custom login pages.

> It will be mobile friendly

This is a very naive statement. Almost all devices connecting to captive portals will be mobile devices. Not many people will bring their desktop computer with them. All mobile devices have built in support for captive portals (see CPD and CPI in the openNDS documentation.).

> I can make changable themes

Yes, see the openNDS documentation.

> with cookies and so on..

Cookies and Javascript are blocked by most mobile devices when displaying the login pages. This is for security reasons to protect the client (guest) device (a part of the functionality of CPD/CPI - built into the device operating system. Again this is not a limitation of openNDS, it is enforced by the client (guest) device vendors/operating system (Apple, Google, Microsoft etc).

> I want to build management system which working for hotels, offices, hospitality and Others.
> For every client - different settings.
> 

This is what openNDS FAS allows you to do. A single Internet hosted FAS server can handle hotels, hospitality etc all at the same time.

> For every client - different settings.

If by "client" you mean the hotel management (in this case) or the owners of a coffee shop, the owners of a bar etc, then yes, no problem, and possible all with the same FAS server.

**Comment by Pok4** [comment]
How to check that my ip is rfc1918? Is there a online tool to do that?
`
I thought you said you will not be charging guests? This might be acceptable to Russians, but guests from EU countries, the UK, the USA, Australia, etc etc will be very upset by this - they will all expect a FREE Internet service.`
Yes i not charge them, I will charge the hotels. 
Every client (guests) will be free of charge. The hotels will be decide that May offer is cheap and good and they can decide about final price for their guest (if the price for room with WiFi access will be higher or smaller) 

The hotels will have statistic how money they must pay in total based on guest Count. (0.50 cents for me per guest) 
They can pay per month or per year (this will be per Contract) 

`This is illegal in the European Union and most other countries. Both you and the hotel would be open to prosecution under EU laws.`

I think that I can  show history of guest Count, total money that they need to pay, bandwidth that is spend in total and something more (if you give me advice what is possible) 


I want to show and live chat like tawk.to in the web.

I see your Answer and when i get the router maybe i cant play with configurations and function. I will read the docs again, but my idea is stuck in my mind.
Im very comfortable with php. I work with mysql too and every data will be stores on mysql db too.
I can set cronjobs for backup, my hosting will make backup too..
I Plan this for a month.

Some guest that not want WiFi can be tolerated in other way. (lowest prices)
I like and one label printer - DYMO LabelWriter 550 and will be integrated in equipment too.
I can use one api to generate WiFi qr codes to login directly + username/password and expire date on same label.
When user is created from portal the button will generate every thing, only staff can choose printer and print the label with data.
I dont think that cost a time. They can open new tab, sessions will be without expiration. They only need to type use and password (or only one code) instead) this can be autogenerate by jacascript or After submit from php. Maybe authorization with codes will be Best option. Its a fast variant.

**Comment by bluewavenet** [comment]
@Pok4 

> How to check that my ip is rfc1918? Is there a online tool to do that?

See section 3 of:
https://www.rfc-editor.org/rfc/rfc1918

ie.:

> The Internet Assigned Numbers Authority (IANA) has reserved the
>    following three blocks of the IP address space for private internets:
> 
>      10.0.0.0        -   10.255.255.255  (10/8 prefix)
>      172.16.0.0      -   172.31.255.255  (172.16/12 prefix)
>      192.168.0.0     -   192.168.255.255 (192.168/16 prefix)

It is not possible for an online tool to show what ip address a router gets from upstream for its WAN interface.
Such a tool will only show the isp's Internet gateway address.

You will have to do something like run  the `ip addr` command on your router to find the WAN interface ip address.
If it is in one of the above rfc1918 address ranges, you will not be able to access the router remotely without a NAT traversal mechanism of some kind. 

OpenNDS solves this almost universal problem by having its own built in NAT traversal mechanism to enable bi-directional FAS/openNDS communications, using https protocol. Https was chosen because all web hosting solutions support https, but very few support ssh.

**Privacy Laws** forbid collection, storage and use of PERSONAL data, (such as a list of web pages visited by someone).

Data such as the total number of client (guest) logins, is not personal, so is acceptable.
OpenNDS already does this. All you would need to do is generate a suitable report for your customer.

> Im very comfortable with php

Good. All the working examples of remote FAS scripts are written in PHP, so you should find it easy to understand.

For example see:
https://github.com/openNDS/openNDS/blob/master/forward_authentication_service/fas-aes/fas-aes-https.php

> I like and one label printer - DYMO LabelWriter 550 and will be integrated in equipment too.
> I can use one api to generate WiFi qr codes to login directly + username/password and expire date on same label.
> When user is created from portal the button will generate every thing, only staff can choose printer and print the label with data.

**_This is far too complicated_** and will have the potential to cause all sorts of problems.

The most common way in use these days is for the remote _FAS to take the user through an automated account generating process_, collecting full name and a valid email address.

Once the user has completed the account generation forms, the user will be allowed a short interval of Internet access, long enough for them to receive a validation email containing a validation link that they must click to continue accessing the Internet. 
I am sure you must have seen this type of process before, it is very common for self creation of accounts on web sites everywhere.

**_Then your customer has to do nothing but pay the amount that is detailed on the report you send them._**

Quite simple for an experienced php programmer to do :-)

**Comment by Pok4** [comment]
Ok, Thanks for your replies.
I will stay with open wrt for sure. Other devices not have such things like this with a lot of documentation..
I read a lot of pdfs Last days and im pretty sure that open wrt is Best. 

The thing about ssh and this limitations for login is worried me. 
I use putty in the past and im able to login in different unix like systems.. In my hosting too.. 
What is this limitations from open wrt? Haha it sounds stupid to me.. 
Anyway, sorry for big lettres, bad autocompletes from my stupid Phone and my bad english. (im not using translators) 
Its a journey, but we Will see the Future whats holds for me :) 
Im closing this one.
